src/common/Utilities/EventMap.h

```diff
@@ -0,0 +1,331 @@
+    * @param phase Phase which should be activated. Values: 1 - 8
+    */
+    void AddPhase(uint8 phase)
+    {
+        if (phase && phase <= 8)
+            _phase |= uint8(1 << (phase - 1));
+    }
+
+    /**
+    * @name RemovePhase
+    * @brief Deactivates the given phase (bitwise).
+    * @param phase Phase which should be deactivated. Values: 1 - 8.
+    */
+    void RemovePhase(uint8 phase)
+    {
+        if (phase && phase <= 8)
+            _phase &= uint8(~(1 << (phase - 1)));
+    }
+
+
+    /**
+    * @name ScheduleEvent
+    * @brief Creates new event entry in map.
+    * @param eventId The id of the new event.
+    * @param time The time in milliseconds as std::chrono::duration until the event occurs.
+    * @param group The group which the event is associated to. Has to be between 1 and 8. 0 means it has no group.
+    * @param phase The phase in which the event can occur. Has to be between 1 and 8. 0 means it can occur in all phases.
+    */
+    void ScheduleEvent(uint32 eventId, Milliseconds const& time, uint32 group = 0, uint8 phase = 0)
+    {
+        ScheduleEvent(eventId, time.count(), group, phase);
+    }
+
+    /**
+    * @name ScheduleEvent
+    * @brief Creates new event entry in map.
+    * @param eventId The id of the new event.
+    * @param time The time in milliseconds until the event occurs.
+    * @param group The group which the event is associated to. Has to be between 1 and 8. 0 means it has no group.
+    * @param phase The phase in which the event can occur. Has to be between 1 and 8. 0 means it can occur in all phases.
+    */
+    void ScheduleEvent(uint32 eventId, uint32 time, uint32 group = 0, uint8 phase = 0);
+
+    /**
+    * @name RescheduleEvent
+    * @brief Cancels the given event and reschedules it.
+    * @param eventId The id of the event.
+    * @param time The time in milliseconds as std::chrono::duration until the event occurs.
+    * @param group The group which the event is associated to. Has to be between 1 and 8. 0 means it has no group.
+    * @param phase The phase in which the event can occur. Has to be between 1 and 8. 0 means it can occur in all phases.
+    */
+    void RescheduleEvent(uint32 eventId, Milliseconds const& time, uint32 group = 0, uint8 phase = 0)
+    {
+        RescheduleEvent(eventId, time.count(), group, phase);
+    }
+
+    /**
+    * @name RescheduleEvent
+    * @brief Cancels the given event and reschedules it.
+    * @param eventId The id of the event.
+    * @param time The time in milliseconds until the event occurs.
+    * @param group The group which the event is associated to. Has to be between 1 and 8. 0 means it has no group.
+    * @param phase The phase in which the event can occur. Has to be between 1 and 8. 0 means it can occur in all phases.
+    */
+    void RescheduleEvent(uint32 eventId, uint32 time, uint32 group = 0, uint8 phase = 0)
+    {
+        CancelEvent(eventId);
+        ScheduleEvent(eventId, time, group, phase);
+    }
+
```

Comment:
```c++
	/**
	* @name RescheduleEvent
	* @brief Cancels the given event and reschedules it.
	* @param eventId The id of the event.
	* @param time The time in milliseconds until the event occurs.
	* @param group The group which the event is associated to. Has to be between 1 and 8. 0 means it has no group.
	* @param phase The phase in which the event can occur. Has to be between 1 and 8. 0 means it can occur in all phases.
	*/
	void RepeatEvent(uint32 time)
	{
		if (Empty())
			return;

		uint32 eventId = _eventMap.begin()->second;
		_eventMap.erase(_eventMap.begin());
		ScheduleEvent(eventId, time);
	}

	/**
	* @name PopEvent
	* @brief Remove the first event in the map.
	*/
	void PopEvent()
	{
		if (!Empty())
			_eventMap.erase(_eventMap.begin());
	}

	/**
	* @name GetEvent
	* @brief Returns the next event to execute.
	* @return Id of the event to execute.
	*/
	uint32 GetEvent()
	{
		while (!Empty())
		{
			EventStore::iterator itr = _eventMap.begin();

			if (itr->first > _time)
				return 0;
			else if (_phase && (itr->second & 0xFF000000) && !(itr->second & (_phase << 24)))
				_eventMap.erase(itr);
			else
				return (itr->second & 0x0000FFFF);
		}

		return 0;
	}
```

Reply:
no just wait when it be finished...